src/parser.rs

```diff
@@ -138,11 +138,20 @@ pub fn tokenize_cla(input: &str) -> Result<Vec<CToken>, ParseError> {
 
 #[doc(hidden)]
 pub fn convert_classic_tokens(tokens: &[CToken]) -> Vec<Token> {
-    _convert_classic_tokens(tokens, &mut Vec::with_capacity(tokens.len()), &mut 0)
+    let num_lambda = tokens.iter().filter(|t| matches!(t, CLambda(_))).count();
+    _convert_classic_tokens(
+        tokens,
+        num_lambda + 1,
+        &mut Vec::with_capacity(tokens.len() - num_lambda),
+        &mut Vec::with_capacity(num_lambda),
```

Comment:
The maximum depth of the stack is less than the number of appearances of the lambda abstract.

The number of free variables is less than the number of tokens minus the number of lambda abstractions.